### DISCUSSION TITLE
How to control colour map of proteins visualized in a Jupyter notebook cell?

### DISCUSSION THREAD

**Discussion by SinaSarparast** [nglviewer/nglview#1037] [Q&A]
Hello,

I'm trying to visualize proteins in a jupyter notebook cell using nglview. By default the colouring theme used is based on the structures. I want to use bfactor as a coloring scheme. I can do that in the nglviewer app from the panel, but in a notebook environment I don't know how to set the right attributes to get the same results as I get when I use the app.
It's very helpful if someone can direct me to the relevant documentation or examples that can help me do so? 

Thanks,

**Comment by hainm** [comment]
hi @SinaSarparast 

Can you do something like `view.add_cartoon(color='bfactor')`?

**Reply by SinaSarparast** [reply]
Hi @hainm ,

Thanks for the answer. I've managed to fix the issue with the following

> view = nv.show_biopython(structure)
> view.remove_cartoon() # to remove previous visualization
> view.add_representation(
>     repr_type="ribbon",
>     colorScheme= "bfactor", colorScale= "spectral", colorReverse= 1
> )
> view

where can I find more information on the `colorScale`? We are replacing prediction errors with the bfactor, so some values can be negative, and we scale it to [-1,1]. Is there a good colorScale that covers negative values as well?

**Reply by hainm** [reply]
> where can I find more information on the `colorScale`? We are replacing prediction errors with the bfactor, so some values can be negative, and we scale it to [-1,1]. Is there a good colorScale that covers negative values as well?

I am not sure below will answer your question, but you can see different value for `colorScale` from the app.

<img width="1339" alt="image" src="https://user-images.githubusercontent.com/4451957/201759535-f920dd5d-0474-4eee-80c0-6f1c1ac03cf6.png">

**Reply by meneshail** [reply]
> Hi @hainm ,
> 
> Thanks for the answer. I've managed to fix the issue with the following
> 
> > view = nv.show_biopython(structure)
> > view.remove_cartoon() # to remove previous visualization
> > view.add_representation(
> > repr_type="ribbon",
> > colorScheme= "bfactor", colorScale= "spectral", colorReverse= 1
> > )
> > view
> 
> where can I find more information on the `colorScale`? We are replacing prediction errors with the bfactor, so some values can be negative, and we scale it to [-1,1]. Is there a good colorScale that covers negative values as well?




Hi,

I had the same question this week and checked the source codes of `ngl`.  It turns out when the `Colormaker` is constructed it will first iterate over each atom to decide the min and max. This will be used as the `domain` parameter later on. So basically your bfactors will be normalized with a min-max scaler to fit in the [0, 1] range that the colormap uses. 

https://github.com/nglviewer/ngl/blob/beb8bf93c001fd36a60a2bf6185f1aa5d503f4d3/src/color/bfactor-colormaker.ts#L34-L50
```typescript
if (!params.domain) {
  let selection
  let min = Infinity
  let max = -Infinity


  if (params.sele) {
    selection = new Selection(params.sele)
  }


  params.structure.eachAtom(function (a) {
    const bfactor = a.bfactor
    min = Math.min(min, bfactor)
    max = Math.max(max, bfactor)
  }, selection)


  this.parameters.domain = [ min, max ]
}
```
